gt/main.c

```diff
@@ -1770,3 +1819,90 @@ run_gt(struct net_config *net_conf, struct gt_config *gt_conf)
 	rte_atomic32_init(&gt_conf->ref_cnt);
 	return 0;
 }
+
+static lua_State *
+alloc_and_setup_lua_state(void)
+{
+	int ret;
+	char lua_entry_path[128];
+	lua_State *lua_state;
+
+	ret = snprintf(lua_entry_path, sizeof(lua_entry_path), \
+			"%s/%s", LUA_POLICY_BASE_DIR, GRANTOR_CONFIG_FILE);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(lua_entry_path));
+
+	lua_state = luaL_newstate();
+	if (lua_state == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gt: failed to create new Lua state at %s!\n",
+			__func__);
+		goto out;
+	}
+
+	luaL_openlibs(lua_state);
+	set_lua_path(lua_state, LUA_POLICY_BASE_DIR);
+	ret = luaL_loadfile(lua_state, lua_entry_path);
+	if (ret != 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gt: %s!\n", lua_tostring(lua_state, -1));
+		goto clean_lua_state;
+	}
+
+	/* Run the loaded chunk. */
+	ret = lua_pcall(lua_state, 0, 0, 0);
+	if (ret != 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gt: %s!\n", lua_tostring(lua_state, -1));
+		goto clean_lua_state;
+	}
+
+	return lua_state;
+
+clean_lua_state:
+	lua_close(lua_state);
+out:
+	return NULL;
+}
+
+void
+update_lua_states(struct gt_config *gt_conf)
+{
+	int i;
+
+	for (i = 0; i < gt_conf->num_lcores; i++) {
+		int ret;
+		struct gt_cmd_entry *entry;
+		struct gt_instance *instance = &gt_conf->instances[i];
+		lua_State *lua_state = alloc_and_setup_lua_state();
+		if (lua_state == NULL) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gt: failed to allocate new lua state to GT block %u\n",
+				i);
+
+			continue;
+		}
+
+		entry = mb_alloc_entry(&instance->mb);
+		if (entry == NULL) {
+			lua_close(lua_state);
+
+			RTE_LOG(WARNING, GATEKEEPER,
```

Comment:
Just for consistency, move the line `lua_close(lua_state);` above immediately before the `continue` below (not seen here).